### PR TITLE
[google_sign_in] Suppress `deprecation` warnings

### PR DESCRIPTION
**Pull request by asashour**
Part of https://github.com/flutter/flutter/issues/91221

There is already a [SuppressWarnings](https://github.com/flutter/plugins/blob/0d5cf74c437e7f922ae27b82bb580e5043f57ceb/packages/google_sign_in/google_sign_in/android/src/main/java/io/flutter/plugins/googlesignin/GoogleSignInPlugin.java#L63)

This PR adds three more, so at least the error message are removed.

I guess removing v1 embedding is something else.

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [X] I signed the [CLA].
- [X] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [X] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by Hixie** [comment]
test-exempt: does not affect logic

**Comment by stuartmorgan** [comment]
Thanks for the submission! Please complete the missing checklist items (in particular, version and CHANGELOG changes, without which this won't pass our CI).

@blasten to review or delegate.

**Review by blasten — CHANGES_REQUESTED** [review]
I'd suggest to remove support entirely instead of the deprecation warnings. cc @GaryQian

**Comment by stuartmorgan** [comment]
I forget, has `flutter` fully dropped support for building with v1? I believe the decision was to keep supporting it in our plugins until that happened.

**Comment by stuartmorgan** [comment]
@GaryQian @blasten Ping on the question about whether we still support v1 on stable.

**Comment by GaryQian** [comment]
V1 apps will still build, but will fail by default and require a flag to ignore.

We plan on removing/breaking v1 in the near future.

**Comment by GaryQian** [comment]
V1 apps will still build, but will fail by default and require a flag to ignore.

We plan on removing/breaking v1 in the near future.

V1 Plugins will still build/remain compatible with apps during the initial v1 deletion, but will call into no-ops eventually in a later phase.

**Comment by blasten** [comment]
> Sounds like this PR (vs removing the code) is correct at the current time then?

Flutter hasn't fully dropped support, but what about removing the offending code? When these APIs are removed, it will be -1 issue.

**Comment by stuartmorgan** [comment]
> > Sounds like this PR (vs removing the code) is correct at the current time then?
> 
> Flutter hasn't fully dropped support, but what about removing the offending code? When these APIs are removed, it will be -1 issue.

See discussion in https://github.com/flutter/plugins/pull/4457

If we're changing the decision about that, we should probably discuss in the linked doc.

File: packages/google_sign_in/google_sign_in/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+## 5.3.1
```
Comment (stuartmorgan) [inline packages/google_sign_in/google_sign_in/CHANGELOG.md:1]: You've updated the version and changelog in the wrong package; the change is to `google_sign_in_android`.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

**Review by stuartmorgan — APPROVED** [review]
LGTM

**Comment by stuartmorgan** [comment]
@blasten Can you re-review? Your change request will be blocking for landing.

**Comment by blasten** [comment]
@stuartmorgan 
I asked permission to see the doc

As for:
> Make relying on v1 support in applications something that is no longer possible, along with whatever tooling support is needed to make that viable (e.g., fatal errors on build if people try).

@GaryQian already added an early tool exit to the Flutter tool with links to the Wiki.

> Make a clear announcement that as of Flutter version X.Y, v1 embedding is non-functional. This should probably be in the release announcement, as well as documented on our wiki for people to easily find.

This is pending.  A breaking change similar to https://docs.flutter.dev/release/breaking-changes/android-v1-embedding-create-deprecation#description-of-change

> Annotate all of the code that this PR removes with a comment saying "Remove this code when this plugin requires a minimum version of X.Y".

I'm not sure this is necessary.  A plugin can drop support to the v1 embedding any time.   One doesn't need to depend on the other.

We mentioned in https://docs.flutter.dev/release/breaking-changes/android-v1-embedding-create-deprecation#description-of-change:

Existing v1 applications consuming a plugin that targets only the v2 embedding won’t build and must migrate. This has been the case since v1.12. However, the likelihood of encountering this increases as plugin developers create and publish v2 only plugins.

It's also technical debt that 1P plugins are using deprecated APIs in some places.  As far as I remember, when v2 was launched, the plugins team did a partial migration to v2 due to time constraints and competing priorities. 

I don't think there's 1 test that can validate that v1 embedding works, and I'm 100% it doesn't work for all newer features introduced after the announcement linked above.

**Comment by GaryQian** [comment]
I will go make the doc public, I didn't realize it wasn't accessible.

We are currently observing metrics on how often builds use v1 embedding and how often the force flag is used to ignore the build failure. We are seeing a consistent downtrend of v1 apps, and v1 apps are a very small percentage of our users now. I haven't formally taken a snapshot of the metrics yet but I can do so and add to the doc. 

One additional incremental step we can take is to remove the force flag just ahead of actual v1 deletion to make it effectively impossible to build without migrating.

**Review by blasten — APPROVED** [review]

**Comment by stuartmorgan** [comment]
> > Make relying on v1 support in applications something that is no longer possible, along with whatever tooling support is needed to make that viable (e.g., fatal errors on build if people try).
> 
> @GaryQian already added an early tool exit to the Flutter tool with links to the Wiki.

But there's a flag to bypass that, as discussed above. And I looked at the warning; it just says v1 "is being deprecated" and
> You may also pass the --ignore-deprecation flag to
> ignore this check and continue with the deprecated v1 embedding. However,
> the v1 Android embedding will be removed in future versions of Flutter.

It doesn't say it's not currently supported. Or that if people use the flag their app still probably won't work correctly, which would be the outcome for most people if all 1P plugins stop supporting v1.

Fundamentally: what is the point of the flag if not to support people continuing to use v1 during this phase of the transition? And if we're supporting people doing that in `flutter` on stable, then we should support it our 1P plugins, because our official, documented policy is to support stable.

> > Annotate all of the code that this PR removes with a comment saying "Remove this code when this plugin requires a minimum version of X.Y".
> 
> I'm not sure this is necessary. A plugin can drop support to the v1 embedding any time. One doesn't need to depend on the other.

A plugin in general *can*, but in general we try to avoid actively breaking people with our plugin updates. (We could make all of our plugins fail to compile on anything but the latest master at any time too, but we don't.)

To clarify, I'm fine with—once X.Y hits stable—a mass PR that removes all the v1 support and bumps all the minimums of the Android implementation packages to X.Y. I'm not saying we need to wait until we would drop pre-X.Y support for other reasons, only that I don't want us to knowingly ship implementations that `pub` will resolve for a pre-X.Y user still using v1, but will actively break their app.

> We mentioned in https://docs.flutter.dev/release/breaking-changes/android-v1-embedding-create-deprecation#description-of-change:
> 
> Existing v1 applications consuming a plugin that targets only the v2 embedding won’t build and must migrate. This has been the case since v1.12. However, the likelihood of encountering this increases as plugin developers create and publish v2 only plugins.

Sure, but there's a substantial difference between "new plugins will start showing up that require v2" and "the Flutter team will actively break you if you are using some of the most popular plugins in the ecosystem, even though they currently work".

> It's also technical debt that 1P plugins are using deprecated APIs in some places. As far as I remember, when v2 was launched, the plugins team did a partial migration to v2 due to time constraints and competing priorities.

Aren't they only "using" them in the sense that they are there for the sole purpose of supporting v1 apps? I.e., if we removed all of the deprecated code, the only effect would be that v1 stopped working? If that's not the case, and we are *missing* v2 codepaths, we should certainly fix that ASAP.

Having support for v1 in flutter/plugins is technical debt for our plugins in exactly the same way that having support for v1 is technical debt for flutter/flutter and flutter/engine. I'm not suggesting that we keep it here forever, only that we be consistent with our support.

**Comment by blasten** [comment]
> But there's a flag to bypass that, as discussed above. And I looked at the warning; it just says v1 "is being deprecated" and

I did a query in Google Analytics, and got that only 0.05% of Gradle builds continued with the `--ignore-deprecation` flag. That is from a 0.29% that received this warning https://github.com/flutter/flutter/blob/496049263e4efd68ced664ee2c90db9c93f08571/packages/flutter_tools/lib/src/project.dart#L599.

I'm not sure if I did this math right, but the data is available for anyone to try it out. I computed this from looking at the total number of Gradle builds, since for some reason the metrics for events related to build apk/appbundle seems wrong.

From a  0.05%, I suspect the number of apps using `google_sign_in` is smaller within that population.

**Comment by stuartmorgan** [comment]
> I did a query in Google Analytics, and got that only 0.05% of Gradle builds continued with the `--ignore-deprecation` flag.

That certainly sounds like a good argument for removing the flag in the next stable release, at which point we could easily do the mass removal I described above, and break *no users* (not even those still using 3.0+v1).

> From a 0.05%, I suspect the % of apps using `google_sign_in` is smaller within that population.

I'm sure that's true. But I don't understand what compelling interest is served by actively breaking even a subset of users who are on a codepath that is *still supported by Flutter stable*, and thus violating the spirit of our documented plugin compatibility commitment, instead of just waiting a few more months.

Is this plugin code a significant maintenance burden? More than the tool and engine code that we were okay with still supporting in 3.0?

**Comment by stuartmorgan** [comment]
Landing as an incremental improvement while we figure out about removal (which should be a discussion that we continue in the doc rather than here).

**Pull request merged**